ext/opcache/Optimizer/sccp.c

```diff
@@ -0,0 +1,1463 @@
+				SET_RESULT(result, &zv);
+				zval_ptr_dtor_nogc(&zv);
+				break;
+			}
+			SET_RESULT_BOT(result);
+			break;
+		case ZEND_ROPE_ADD:
+		case ZEND_ROPE_END:
+			// TODO The way this is currently implemented will result in quadratic runtime
+			// This is not necessary, the way the algorithm works it's okay to reuse the same
+			// string for all SSA vars with some extra checks
+			SKIP_IF_TOP(op1);
+			SKIP_IF_TOP(op2);
+			if (zend_optimizer_eval_binary_op(&zv, ZEND_CONCAT, op1, op2) == SUCCESS) {
+				SET_RESULT(result, &zv);
+				zval_ptr_dtor_nogc(&zv);
+				break;
+			}
+			SET_RESULT_BOT(result);
+			break;
+		case ZEND_INIT_ARRAY:
+		case ZEND_ADD_ARRAY_ELEMENT:
+		{
+			zval *result = NULL;
+			if (opline->extended_value & ZEND_ARRAY_ELEMENT_REF) {
+				SET_RESULT_BOT(result);
+				SET_RESULT_BOT(op1);
+				break;
+			}
+
+			if (opline->opcode == ZEND_ADD_ARRAY_ELEMENT) {
+				result = &ctx->values[ssa_op->result_use];
+				if (IS_BOT(result)) {
+					SET_RESULT_BOT(result);
+					break;
+				}
+				SKIP_IF_TOP(result);
+			}
+
+			SKIP_IF_TOP(op1);
+			if (op2) {
+				SKIP_IF_TOP(op2);
+			}
+
+			/* We want to avoid keeping around intermediate arrays for each SSA variable in the
+			 * ADD_ARRAY_ELEMENT chain. We do this by only keeping the array on the last opcode
+			 * and use a NULL value everywhere else. */
+			if (Z_TYPE(ctx->values[ssa_op->result_def]) == IS_NULL) {
+				break;
+			}
+
+			if (result) {
+				ZVAL_COPY_VALUE(&zv, result);
+				ZVAL_NULL(result);
+			} else {
+				array_init(&zv);
+			}
+
+			if (ct_eval_add_array_elem(&zv, op1, op2) == SUCCESS) {
+				SET_RESULT(result, &zv);
+				zval_ptr_dtor_nogc(&zv);
+				break;
+			}
+			SET_RESULT_BOT(result);
+			zval_ptr_dtor_nogc(&zv);
+			break;
+		}
+		case ZEND_ASSIGN_DIM:
+		{
+			zval *data = get_op1_value(ctx, opline+1, ssa_op+1);
+			if (IS_BOT(data)) {
+				SET_RESULT_BOT(op1);
+				SET_RESULT_BOT(result);
+				break;
+			}
+
+			SKIP_IF_TOP(data);
+			SKIP_IF_TOP(op1);
+			if (op2) {
+				SKIP_IF_TOP(op2);
+			}
+
+			ZVAL_DUP(&zv, op1);
+			if (ct_eval_assign_dim(&zv, data, op2) == SUCCESS) {
+				SET_RESULT(result, data);
+				SET_RESULT(op1, &zv);
+				zval_ptr_dtor_nogc(&zv);
+				break;
+			}
+			SET_RESULT_BOT(result);
+			SET_RESULT_BOT(op1);
+			zval_ptr_dtor_nogc(&zv);
+			break;
+		}
+		case ZEND_DO_ICALL:
+		{
+			zend_call_info *call;
+			zval *name, *args[2] = {NULL};
+			int i;
+
+			if (!ctx->call_map) {
+				SET_RESULT_BOT(result);
+				break;
+			}
+
+			call = ctx->call_map[opline - ctx->scdf.op_array->opcodes];
+			name = CT_CONSTANT_EX(ctx->scdf.op_array, call->caller_init_opline->op2.constant);
+
+			/* We already know it can't be evaluated, don't bother checking again */
+			if (ssa_op->result_def < 0 || IS_BOT(&ctx->values[ssa_op->result_def])) {
+				break;
+			}
+
+			/* We're only interested in functions with one or two arguments right now */
+			if (call->num_args == 0 || call->num_args > 2) {
```

Comment:
The 3-arg `version_compare` check in eval_func_call won't be hit with this constraint.